File: Python/Dynamic Programming/L_houseRobber.py

from functools import cache
from typing import List


def rob(nums) -> int:
        n = len(nums)
        @cache
        def robs(i):
            if i >= n:
                return 0
            return nums[i] + max(robs(i+2), robs(i+3))
        return max(robs(0), robs(1))

def rob1(nums) -> int:
        if len(nums) == 1:
            return nums[0]
        first = nums[0]
        second = max(nums[1], nums[0])
        for i in range(2, len(nums)):
            curr = max(second, first + nums[i])
            first = second
            second = curr
        return second
    
def rob2(nums: List[int]) -> int:
    prev1, prev2 = 0, 0
    for i in nums:
        prev1, prev2 = prev2, max(prev1+i, prev2)
    return prev2

File: cpp/Dynamic Programming/L_houseRobber.cpp
#include <bits/stdc++.h>
using namespace std;

int rob(vector<int> &nums)
{
    int prev = 0;
    int curr = 0;
    int next = 0;

    for (int i = 0; i < nums.size(); i++)
    {
        next = max(prev + nums[i], curr);
        prev = curr;
        curr = next;
    }

    return curr;
}

int rob1(vector<int> &nums)
{
    int n = nums.size();
    vector<int> dp(n, -1);
    if (n == 0)
        return 0;
    if (n == 1)
        return nums[0];

    int ans = helper(n - 1, nums, dp);
    return ans;
}

int helper(int ind, vector<int> nums, vector<int> &dp)
{
    if (ind == 0)
        return nums[ind];
    if (ind == -1)
        return 0;
    if (dp[ind] != -1)
        return dp[ind];
    int pick = nums[ind] + helper(ind - 2, nums, dp);
    int non_pick = 0 + helper(ind - 1, nums, dp);
    return dp[ind] = max(pick, non_pick);
}